chain/index/chain-indexing-overview-for-rpc-providers.MD

```diff
@@ -0,0 +1,218 @@
+[ChainIndexer]
+  GCRetentionEpochs = X  # Replace X with your desired value
+```
+
+- Setting `GCRetentionEpochs` to 0 (**default**) disables GC.
+- Any positive value enables GC and determines the number of epochs of historical data to retain.
+
+#### Recommendations
+
+1. **Archival Nodes**: **Keep GC disabled** (`GCRetentionEpochs` = 0) to retain all indexed data.
+
+2. **Non-Archival Nodes**:  Set `GCRetentionEpochs` to match the amount of chain state your node retains 
+(*Example:* if your node is configured to retain 2 days of Filecoin chain state with the Splitstore, set `GCRetentionEpochs` to `retentionDays * epochsPerDay = 2 * 2880 = 5760`).
+**Warning:** Setting this value below the chain state retention period will degrade RPC performance and reliability because the Index will lack data for epochs still present in the chain state.  
+
+### Removed Options
+
+**Note: The following config options no longer exist in Lotus and have been removed in favor of the ChainIndexer config options explained above. They can be removed when upgrading to Lotus v1.31.0.**
+
+```toml
+[Fevm]
+EthTxHashMappingLifetimeDays = 0
+
+[Events]
+DisableHistoricFilterAPI = false
+DatabasePath = ""
+```
+
+## Upgrade Steps
+
+> **Note:** One can upgrade/downgrade between [pre-ChainIndexer](#previous-indexing-system) and [with-ChainIndexer](#chainindexer-indexing-system) Lotus versions without conflict because they persist state to different directories and don't rely on each other. No backup is necessary (but extra backups don't hurt).  
+
+Upgrading to the new `ChainIndexer` involves these steps:
+
+1. **Stop the Lotus Node**
+   - Stop your Lotus node before starting the upgrade and migration process.
+2. **Update Configuration**
+   - Modify your Lotus configuration to enable the `ChainIndexer` as described in the [`ChainIndexer Config` section above](#chainindexer-config).
+3. **Restart Lotus Node**
+   - Restart your Lotus node with the new configuration.
+   - The `ChainIndexer` will begin indexing **real-time chain state changes** immediately in the `${LOTUS_PATH}/chainindex` directory.
+   - **However, it will not automatically index any historical chain state (i.e., any previously existing chain state). To perform backfilling, please see the [`Backfill` section below](#backfill).**
+
+> **Note:** It's recommended to keep the [pre-ChainIndexer](#previous-indexing-system) state directory around until you've confirmed you don't need to [downgrade](#downgrade).  After sustained successful operations after the upgrade, the [pre-ChainIndexer](#previous-indexing-system) state directory can be removed to reclaim disk space.  
+
+## Backfill
+There is no migration from [pre-ChainIndexer indices](#previous-indexing-system) to the [ChainIndex](#chainindexer-indexing-system) because the [pre-ChainIndexer indices](#previous-indexing-system) are not fully reliable.  Instead one needs to index historical chain state (i.e., backfill) using one of the following tools.
```

Comment:
```suggestion
There is no automated migration from [pre-ChainIndexer indices](#previous-indexing-system) to the [ChainIndex](#chainindexer-indexing-system) because the [pre-ChainIndexer indices](#previous-indexing-system) are not fully reliable.  Instead one needs to index historical chain state (i.e., backfill) using one of the following tools.
```

Reply:
I think this was my addition.  I've been assuming it's not possible to go from pre-ChainIndexer 3 dbs to ChainIndexer DB because we don't have certainty that previous system is void of holes.  If we think it would be possible to bootstrap ChainIndexer from the previous DBs, we may want to look into that because I assume that would drastically reduce the upgrade time for archival nodes.  (Thread on this: https://github.com/filecoin-project/lotus/issues/12453#issuecomment-2403463045 )

Reply:
1) We don't have the confidence that the old Indexing system/DBs have the correct data. We've fixed multiple bugs in the old Indexing system over the past few months but we still see report of holes/missing data/mismatched event index counts etc. See recent complaints by Oku and Forrest as examples.  I do not feel confident bootstrapping the new Index from the old one because this is our chance to get it right.

2) This backfilling is a one time cost. If an RPC provider is running multiple nodes, they only need to do it on one node and can then simply copy over the Index. The new Index DB copy can also be shared among RPC providers if there is a trust relationship.

3) The schemas of the old fragmented Indices dont's easily/natuarally map to the schema of the new Index. There will be additional work/data wrangling to get this right.

I would like to make a strong push back against the idea of using the old Indices for filling up the new one just to avoid having to pay a one time cost per RPC provider given the low degree of confidence in the consistency/correctness in the old Indices. That lack of consistency/correctness in the fragmented DBs is what led us to doing this workstream in the first place.

Reply:
Also -> this is only for Archival nodes not for snapshot synced nodes. For snapshot synced nodes with only a portion of the chain state (who only serve queries going back a few days) -> it will be much faster.  See https://filecoinproject.slack.com/archives/CP50PPW2X/p1727862474601319.

Reply:
I'll bring this into https://github.com/filecoin-project/lotus/pull/12600